### ISSUE TITLE
sql no rows in result set on lxc copy

### ISSUE THREAD

**Issue by cryptofuture** [lxc/lxd#3137]
`lxc copy target:con6 con6` fails with
```
error: sql: no rows in result set
```
   >Note: I already recreated (btw, looks like we still have no copy in lxc network)  lxdbr2 interface and copied limit profile from target

Ubuntu Xenial 16.04.x
[monitor-log.txt](https://github.com/lxc/lxd/files/886466/monitor-log.txt)
[copy-debug.txt](https://github.com/lxc/lxd/files/886461/copy-debug.txt)

```
:lxc info
  driver: lxc
  driver_version: 2.0.7
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.4.0-71-generic
  server: lxd
  server_pid: 944
  server_version: "2.12"
  storage: btrfs
  storage_version: "4.4"
```
```
:target lxc info
  driver: lxc
  driver_version: 2.0.7
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.10.0-13-generic
  server: lxd
  server_pid: 4590
  server_version: "2.12"
  storage: zfs
  storage_version: 0.6.5.9-2
```

```
:lxc profile show limit
config: {}
description: ""
devices:
  eth0:
    limits.egress: 600Kbit
    limits.ingress: 600Kbit
    nictype: bridged
    parent: lxdbr2
    type: nic
  root:
    path: /
    pool: lxd
    type: disk
name: limit
used_by:
- /1.0/containers/con6
```
```
:lxc config show con6
architecture: x86_64
config:
  volatile.base_image: id
  volatile.eth0.hwaddr: mac
  volatile.eth0.name: eth0
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":100000,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":100000,"Nsid":0,"Maprange":65536}]'
  volatile.last_state.power: STOPPED
  volatile.root.hwaddr: mac
  volatile.root.name: eth1
devices:
  root:
    path: /
    pool: lxd
    type: disk
ephemeral: false
profiles:
- limit
```

**Comment by stgraber** [comment]
Do you have a storage pool called "lxd" on the target server?

**Comment by cryptofuture** [comment]
Oh, I understand what you suggesting, and most likely it is. lxd storage still remains the same in copied profile, but i have no `lxd` storage on another machine, its could `default` on it.
Will try to edit profile.

**Comment by stgraber** [comment]
Note that the error message totally needs to be fixed as it's very confusing :)

But since your source container has a direct device entry for a root using the "lxd" pool, LXD will fail to re-create the container on the target unless there is a similarly named pool available.

The best option may be to remove that container device entry, letting it inherit from the default profile instead, then migrate it again, which would have LXD use the profile of the remote server to select what storage pool to use.

**Comment by stgraber** [comment]
@brauner can you reproduce this issue and fix the error to make a bit more sense to users? :)

**Comment by cryptofuture** [comment]
Btw, I still can test myself. Copying 60Gb container right now, more ~80% done. So I can try again only after it will finish.

**Comment by cryptofuture** [comment]
Thats what I have after profile editing now, still same error.
```
config: {}
description: ""
devices:
  eth0:
    limits.egress: 600Kbit
    limits.ingress: 600Kbit
    nictype: bridged
    parent: lxdbr2
    type: nic
  root:
    path: /
    pool: default
    type: disk
name: limit
used_by: []
```
```
+---------+--------+----------------------------+---------+
|  NAME   | DRIVER |           SOURCE           | USED BY |
+---------+--------+----------------------------+---------+
| default | btrfs  | /var/lib/lxd/disks/default | 31      |
+---------+--------+----------------------------+---------+
```

**Comment by stgraber** [comment]
@cryptofuture right, but based on what you posted before, your container itself overrides that "root" device so having it in the profile won't do you much good.

You'd need to do:

    lxc config device remove con6 root

Which will then have it inherit the "root" device from your "limit" profile.

At which point, copying the container to the target server should work, so long as the target server's "limit" profile also contains a valid "root" device.

**Comment by cryptofuture** [comment]
Sorry, I gave incorrect info.
I tried another container with the same error, and another custom profile. After fixing pool name on it too, lxc copy it works. And it works for both containers with same error.
**Overall:** 
_Copying container:location <-> container always works and changes pool name to actual pool name on destination.  
But copying profiles, saves the same pool name (in profile), and it breaks container copy._

So this is not exactly the bug, but will be cool to make some enchantment to profile copy, especially, when single pool active on destination.
Or better lxc copy should copy/recopy custom profile with container every time, and rename poolname to actual poolname on destination (maybe harder to implement).

**Comment by stgraber** [comment]
Hmm, so you were able to copy a profile with an invalid "pool" property? That sounds like a bug. The "lxc profile copy" should fail in such cases.

**Comment by stgraber** [comment]
Hmm, indeed we seem to be able to set the "pool" property to whatever value we want, that's no good.

**Comment by cryptofuture** [comment]
>Hmm, so you were able to copy a profile with an invalid "pool" property? That sounds like a bug. The "lxc profile copy" should fail in such cases.

lxc profile copy just creates totally the same profile on destination, at least looks like it do, for me.

**Comment by stgraber** [comment]
yeah, I'm fixing that code now so that it fails if pointing to a storage pool which doesn't exist.

**Comment by cryptofuture** [comment]
I think basically most old lxd version used `lxd` poolname by default, and not much people use non default profiles, so we find out that a bit late. But basically when I want give some limits, settings, different networks to another 15 containers or bigger group, custom profiles really easier to manage, than just editing every container.

**Comment by stgraber** [comment]
Storage pools are very new (introduced in 2.9) and I think the default name varies based on the kind of storage you had defined before. If storage.zfs_pool_name was set to "lxd", then the pool will be called "lxd", for just about every other case, it'll be called "default", causing the difference you're seeing.

**Comment by stgraber** [comment]
The branch I just sent will prevent profiles from being created or edited to point at non-existent storage pools. This should prevent the error you were getting from showing up in the first place.

**Comment by cryptofuture** [comment]
Yes, I see changes, really thanks for it. And this is more easy solution, than the one I was thinking about.
lxc network copy not exist for the "same" reason? Since usually user wants new subnet, but the same interface name (and copying just he same configuration may break routing) Or thats just because lxc network still quite young?

**Comment by stgraber** [comment]
Yeah, we've not felt a strong need for "lxc network copy" or "lxc storage copy" as both network and storage objects usually include host-specific configuration making them unlikely candidates for copies.
